core/include/userver/engine/io/common.hpp

```diff
@@ -51,6 +52,12 @@ class ReadableBase {
   impl::ContextAccessor* ca_{nullptr};
 };
 
+class NonblockingReadableBase {
+ public:
+  // Either receives len bytes or returns -1 AND EAGAIN/EWOULDBLOCK
+  [[nodiscard]] virtual int64_t ReadNonblocking(void* buf, size_t len) = 0;
```

Comment:
`std::size_t TryReadSome`

Reply:
I think I'm unable to do so - currently all possible size_t values are handled already (e.g. 0 has semantics RST/FIN, >0 has semantics 'data received').
To be consistent we need to shift to int64_t as ssize_t (https://man.freebsd.org/cgi/man.cgi?query=recv(2)) alternative.

Also, I wonder if it's good idea to map ssize_t to size_t and cloak negative values in internal core interfaces..

Reply:
ASIO treats such cases as errors (it is reported as `boost::asio::error::eof`). As I see, we already report FIN as an error in ReadSome https://github.com/userver-framework/userver/blob/b5df56fee6feebbc285e521abf8d7316dff7001e/core/src/engine/io/fd_control.hpp#L175

How about returning `std::size_t` with 0 if there's no data at the moment, and report connection FIN by throwing an exception?

Reply:
Just to mention, I think I've proposed the more clean iface for it doesn't cloak real retvals, but in contrary you're dropping hints that we'd better do it in 'existing' way and use current ifaces...  Makes sense.. however:

>How about returning std::size_t with 0 if there's no data at the moment, and report connection FIN by throwing an exception?

Actually, it was my first attempt to do this PR, but time went and now I don't like the idea. But u're maintainer and the dicision is up to you anyway. But! can we discuss the use case one more time?
When we speak in terms ReadNonblocking-iface, I think, any behavior is logically-possible. 
When we speak about existing ~ReadableBase~ iface - currently it's Read()-methods have yet another semantics. They don't return 0 and don't behave this way, so imo it kinda do the iface less... symmetrical? consistent?

So the proposal was to 1) don't cloak 2) do the iface which doesn't have known semantics so far. My point is if we'd left that in separate iface, the contract wouldn't have had this flaws. But in ReadableBase? idk... looks as a hard decision  :) to me. 
I would appreciate any clarifications. And I definitely can do as you wish nevertheless.

Reply:
We'll discuss the API tomorrow with the whole team. I'll come back with a more weighted opinion soon

Reply:
How about `std::optional<std::size_t> TryReadSome(void* buf, size_t len)`

If there's a need to check for data and ignore failures, then write something like: 
```
if (s.TryReadSome(buf, size).value_or(0)) // throws nothing
```

To distinguish errors and no-data-ready:
```
auto received = s.TryReadSome(buf, size);
if (!received) ByeBye();
if (*received != 0) ProcessData(buf, *received);
```
What do you think?

Reply:
>How about [[nodiscard]] std::optional<std::size_t> TryReadSome(void* buf, std::size_t len)

ReadSome reads at least 1 byte. ReadAll reads all bytes. TryReadSome() vs TryReadAll()? 
I hope naming like ```ReadAny``` would fit.

> ```std::optional<std::size_t>```

I think you're very persistent to not use int64_t/ssize_t (the latter is hardly to use for it's more likely related to socket API) 
 :) .
Strictly speaking, I think it would fit also.
Very well, I'll shift to this optional thing (I hope I'll manage today. Or in ~3 days otherwise). 
Thanks for your time.

Reply:
Just add `UINVARIANT(false, "not implemented");` if the implementation is not trivial. It's better to explicitly fail